crates/sncast/src/starknet_commands/script.rs

```diff
@@ -38,14 +40,182 @@ use sncast::helpers::configuration::CastConfig;
 use sncast::helpers::constants::SCRIPT_LIB_ARTIFACT_NAME;
 use sncast::response::print::print_as_warning;
 use sncast::response::structs::ScriptResponse;
+use sncast::{ErrorData, TransactionError};
 use starknet::accounts::Account;
+use starknet::core::types::StarknetError;
 use starknet::core::types::{BlockId, BlockTag::Pending, FieldElement};
 use starknet::providers::jsonrpc::HttpTransport;
-use starknet::providers::JsonRpcClient;
+use starknet::providers::{JsonRpcClient, ProviderError};
 use tokio::runtime::Runtime;
 
 type ScriptStarknetContractArtifacts = StarknetContractArtifacts;
 
+pub trait SerializeAsFelt252 {
+    fn serialize_as_felt252(&self) -> Vec<Felt252>;
+}
+
+#[derive(Debug)]
+enum ScriptCommandError {
+    ContractArtifactsNotFound(ErrorData),
+    ScriptProviderError(ScriptProviderError),
+}
+
+impl From<RecoverableStarknetCommandError> for ScriptCommandError {
+    fn from(value: RecoverableStarknetCommandError) -> Self {
+        match value {
+            RecoverableStarknetCommandError::ContractArtifactsNotFound(err_data) => {
+                ScriptCommandError::ContractArtifactsNotFound(err_data)
+            }
+            RecoverableStarknetCommandError::TransactionError(error) => match error {
+                TransactionError::Rejected => ScriptCommandError::ScriptProviderError(
+                    ScriptProviderError::StarknetError(ScriptStarknetError::TransactionRejected),
+                ),
+                TransactionError::Reverted(err_data) => {
+                    ScriptCommandError::ScriptProviderError(ScriptProviderError::StarknetError(
+                        ScriptStarknetError::TransactionReverted(err_data),
+                    ))
+                }
+            },
+            RecoverableStarknetCommandError::ProviderError(error) => {
+                ScriptCommandError::ScriptProviderError(error.into())
+            }
+        }
+    }
+}
+
+impl SerializeAsFelt252 for ScriptCommandError {
+    fn serialize_as_felt252(&self) -> Vec<Felt252> {
+        match self {
+            ScriptCommandError::ContractArtifactsNotFound(err) => {
+                let mut res = vec![Felt252::from(0)];
+                res.extend(ByteArray::from(err.data.clone()).serialize());
+                res
+            }
+            ScriptCommandError::ScriptProviderError(err) => {
+                let mut res = vec![Felt252::from(1)];
+                res.extend(err.serialize_as_felt252());
+                res
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+enum ScriptProviderError {
+    Other(ErrorData),
```

Comment:
technically, what are possible errors behind "other"?

Reply:
https://github.com/xJonathanLEI/starknet-rs/blob/3b89b739351b5a2700d6bc98ec91694e6fbcceb8/starknet-providers/src/provider.rs#L306-L316

from starknet-rs:
```
pub enum ProviderError {
    #[error(transparent)]
    StarknetError(StarknetError),
    #[error("Request rate limited")]
    RateLimited,
    #[error("Array length mismatch")]
    ArrayLengthMismatch,
    #[error("{0}")]
    Other(Box<dyn ProviderImplError>),
}
```
`ProviderError::ArrayLengthMismatch` and `ProviderError::Other` are mapped to `ScriptProviderError::Other`

Reply:
ok thanks. Does it make sense then to rename it on our end to `ProviderError` ? It may be a bit more descriptive

Reply:
okay

Reply:
https://github.com/foundry-rs/starknet-foundry/pull/1705/commits/7669401f5f9a95af78ccfc2a930d7f1cc6049914